File: dsa-interview/java/Trie.java
package java;

/**
 * Trie (Prefix Tree) Implementation
 * 
 * Time Complexity:
 * - Insert: O(m) where m is the length of the word
 * - Search: O(m)
 * - StartsWith: O(m)
 * - Delete: O(m)
 * 
 * Space Complexity: O(ALPHABET_SIZE * m * n) where n is number of words
 * 
 * Common Interview Problems:
 * - Autocomplete/Search Suggestions
 * - Word Search II (Leetcode 212)
 * - Implement Trie (Leetcode 208)
 * - Word Dictionary with wildcards (Leetcode 211)
 * - Maximum XOR of Two Numbers (Leetcode 421)
 */
public class Trie {
    
    private static final int ALPHABET_SIZE = 26;
    
    private class TrieNode {
        TrieNode[] children;
        boolean isEndOfWord;
        int prefixCount;  // Count of words with this prefix
        int wordCount;    // Count of this exact word (for duplicates)
        
        TrieNode() {
            children = new TrieNode[ALPHABET_SIZE];
            isEndOfWord = false;
            prefixCount = 0;
            wordCount = 0;
        }
    }
    
    private final TrieNode root;
    
    public Trie() {
        root = new TrieNode();
    }
    
    /**
     * Inserts a word into the trie.
     */
    public void insert(String word) {
        if (word == null || word.isEmpty()) return;
        
        TrieNode current = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (current.children[index] == null) {
                current.children[index] = new TrieNode();
            }
            current = current.children[index];
            current.prefixCount++;
        }
        current.isEndOfWord = true;
        current.wordCount++;
    }
    
    /**
     * Returns true if the word is in the trie.
     */
    public boolean search(String word) {
        TrieNode node = searchNode(word);
        return node != null && node.isEndOfWord;
    }
    
    /**
     * Returns true if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(String prefix) {
        return searchNode(prefix) != null;
    }
    
    /**
     * Returns count of words with given prefix.
     */
    public int countWordsWithPrefix(String prefix) {
        TrieNode node = searchNode(prefix);
        return node == null ? 0 : node.prefixCount;
    }
    
    /**
     * Returns count of exact word occurrences.
     */
    public int countExactWord(String word) {
        TrieNode node = searchNode(word);
        return node == null ? 0 : node.wordCount;
    }
    
    /**
     * Deletes a word from the trie.
     */
    public boolean delete(String word) {
        if (!search(word)) return false;
        deleteHelper(root, word, 0);
        return true;
    }
    
    private boolean deleteHelper(TrieNode current, String word, int index) {
        if (index == word.length()) {
            if (!current.isEndOfWord) return false;
            current.isEndOfWord = false;
            current.wordCount--;
            return isEmpty(current);
        }
        
        int charIndex = word.charAt(index) - 'a';
        TrieNode child = current.children[charIndex];
        if (child == null) return false;
        
        child.prefixCount--;
        boolean shouldDeleteChild = deleteHelper(child, word, index + 1);
        
        if (shouldDeleteChild) {
            current.children[charIndex] = null;
            return isEmpty(current) && !current.isEndOfWord;
        }
        return false;
    }
    
    private boolean isEmpty(TrieNode node) {
        for (TrieNode child : node.children) {
            if (child != null) return false;
        }
        return true;
    }
    
    private TrieNode searchNode(String word) {
        if (word == null) return null;
        
        TrieNode current = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (current.children[index] == null) {
                return null;
            }
            current = current.children[index];
        }
        return current;
    }
    
    /**
     * Search with wildcard '.' that can match any character.
     * Used in: Design Add and Search Words Data Structure (Leetcode 211)
     */
    public boolean searchWithWildcard(String word) {
        return searchWithWildcardHelper(root, word, 0);
    }
    
    private boolean searchWithWildcardHelper(TrieNode node, String word, int index) {
        if (index == word.length()) {
            return node.isEndOfWord;
        }
        
        char c = word.charAt(index);
        if (c == '.') {
            // Try all possible characters
            for (TrieNode child : node.children) {
                if (child != null && searchWithWildcardHelper(child, word, index + 1)) {
                    return true;
                }
            }
            return false;
        } else {
            int charIndex = c - 'a';
            if (node.children[charIndex] == null) {
                return false;
            }
            return searchWithWildcardHelper(node.children[charIndex], word, index + 1);
        }
    }
    
    /**
     * Get all words with given prefix (for autocomplete).
     */
    public java.util.List<String> getWordsWithPrefix(String prefix) {
        java.util.List<String> result = new java.util.ArrayList<>();
        TrieNode node = searchNode(prefix);
        if (node != null) {
            collectWords(node, new StringBuilder(prefix), result);
        }
        return result;
    }
    
    private void collectWords(TrieNode node, StringBuilder prefix, java.util.List<String> result) {
        if (node.isEndOfWord) {
            result.add(prefix.toString());
        }
        for (int i = 0; i < ALPHABET_SIZE; i++) {
            if (node.children[i] != null) {
                prefix.append((char) ('a' + i));
                collectWords(node.children[i], prefix, result);
                prefix.deleteCharAt(prefix.length() - 1);
            }
        }
    }
    
    // ==================== DEMO ====================
    public static void main(String[] args) {
        Trie trie = new Trie();
        
        // Basic operations
        trie.insert("apple");
        trie.insert("app");
        trie.insert("application");
        trie.insert("apply");
        
        System.out.println("Search 'apple': " + trie.search("apple"));       // true
        System.out.println("Search 'app': " + trie.search("app"));           // true
        System.out.println("Search 'appl': " + trie.search("appl"));         // false
        System.out.println("StartsWith 'app': " + trie.startsWith("app"));   // true
        System.out.println("Words with prefix 'app': " + trie.countWordsWithPrefix("app")); // 4
        
        System.out.println("All words with prefix 'app': " + trie.getWordsWithPrefix("app"));
        
        // Wildcard search
        trie.insert("bad");
        trie.insert("dad");
        trie.insert("mad");
        System.out.println("Search '.ad': " + trie.searchWithWildcard(".ad")); // true
        System.out.println("Search 'b..': " + trie.searchWithWildcard("b..")); // true
        
        // Delete
        trie.delete("apple");
        System.out.println("After delete 'apple', search 'apple': " + trie.search("apple")); // false
        System.out.println("After delete 'apple', search 'app': " + trie.search("app"));     // true
    }
}

File: dsa-interview/cpp/Trie.cpp
/**
 * Trie (Prefix Tree) Implementation in C++
 * 
 * Time Complexity:
 * - Insert: O(m) where m is the length of the word
 * - Search: O(m)
 * - StartsWith: O(m)
 * - Delete: O(m)
 * 
 * Space Complexity: O(ALPHABET_SIZE * m * n) where n is number of words
 */

#include <iostream>
#include <string>
#include <vector>
#include <memory>

class Trie {
private:
    static const int ALPHABET_SIZE = 26;
    
    struct TrieNode {
        std::unique_ptr<TrieNode> children[ALPHABET_SIZE];
        bool isEndOfWord;
        int prefixCount;
        int wordCount;
        
        TrieNode() : isEndOfWord(false), prefixCount(0), wordCount(0) {
            for (int i = 0; i < ALPHABET_SIZE; i++) {
                children[i] = nullptr;
            }
        }
    };
    
    std::unique_ptr<TrieNode> root;
    
    TrieNode* searchNode(const std::string& word) {
        TrieNode* current = root.get();
        for (char c : word) {
            int index = c - 'a';
            if (!current->children[index]) {
                return nullptr;
            }
            current = current->children[index].get();
        }
        return current;
    }
    
    bool isEmpty(TrieNode* node) {
        for (int i = 0; i < ALPHABET_SIZE; i++) {
            if (node->children[i]) return false;
        }
        return true;
    }
    
    bool deleteHelper(TrieNode* current, const std::string& word, int index) {
        if (index == word.length()) {
            if (!current->isEndOfWord) return false;
            current->isEndOfWord = false;
            current->wordCount--;
            return isEmpty(current);
        }
        
        int charIndex = word[index] - 'a';
        if (!current->children[charIndex]) return false;
        
        current->children[charIndex]->prefixCount--;
        bool shouldDeleteChild = deleteHelper(current->children[charIndex].get(), word, index + 1);
        
        if (shouldDeleteChild) {
            current->children[charIndex].reset();
            return isEmpty(current) && !current->isEndOfWord;
        }
        return false;
    }
    
    bool searchWithWildcardHelper(TrieNode* node, const std::string& word, int index) {
        if (index == word.length()) {
            return node->isEndOfWord;
        }
        
        char c = word[index];
        if (c == '.') {
            for (int i = 0; i < ALPHABET_SIZE; i++) {
                if (node->children[i] && 
                    searchWithWildcardHelper(node->children[i].get(), word, index + 1)) {
                    return true;
                }
            }
            return false;
        } else {
            int charIndex = c - 'a';
            if (!node->children[charIndex]) return false;
            return searchWithWildcardHelper(node->children[charIndex].get(), word, index + 1);
        }
    }
    
    void collectWords(TrieNode* node, std::string& prefix, std::vector<std::string>& result) {
        if (node->isEndOfWord) {
            result.push_back(prefix);
        }
        for (int i = 0; i < ALPHABET_SIZE; i++) {
            if (node->children[i]) {
                prefix.push_back('a' + i);
                collectWords(node->children[i].get(), prefix, result);
                prefix.pop_back();
            }
        }
    }

public:
    Trie() : root(std::make_unique<TrieNode>()) {}
    
    void insert(const std::string& word) {
        if (word.empty()) return;
        
        TrieNode* current = root.get();
        for (char c : word) {
            int index = c - 'a';
            if (!current->children[index]) {
                current->children[index] = std::make_unique<TrieNode>();
            }
            current = current->children[index].get();
            current->prefixCount++;
        }
        current->isEndOfWord = true;
        current->wordCount++;
    }
    
    bool search(const std::string& word) {
        TrieNode* node = searchNode(word);
        return node != nullptr && node->isEndOfWord;
    }
    
    bool startsWith(const std::string& prefix) {
        return searchNode(prefix) != nullptr;
    }
    
    int countWordsWithPrefix(const std::string& prefix) {
        TrieNode* node = searchNode(prefix);
        return node ? node->prefixCount : 0;
    }
    
    int countExactWord(const std::string& word) {
        TrieNode* node = searchNode(word);
        return node ? node->wordCount : 0;
    }
    
    bool deleteWord(const std::string& word) {
        if (!search(word)) return false;
        deleteHelper(root.get(), word, 0);
        return true;
    }
    
    bool searchWithWildcard(const std::string& word) {
        return searchWithWildcardHelper(root.get(), word, 0);
    }
    
    std::vector<std::string> getWordsWithPrefix(const std::string& prefix) {
        std::vector<std::string> result;
        TrieNode* node = searchNode(prefix);
        if (node) {
            std::string current = prefix;
            collectWords(node, current, result);
        }
        return result;
    }
};

// ==================== XOR Trie for Maximum XOR Problems ====================
class XORTrie {
private:
    static const int MAX_BITS = 31;
    
    struct TrieNode {
        std::unique_ptr<TrieNode> children[2];
        int count;
        
        TrieNode() : count(0) {
            children[0] = nullptr;
            children[1] = nullptr;
        }
    };
    
    std::unique_ptr<TrieNode> root;

public:
    XORTrie() : root(std::make_unique<TrieNode>()) {}
    
    void insert(int num) {
        TrieNode* current = root.get();
        for (int i = MAX_BITS; i >= 0; i--) {
            int bit = (num >> i) & 1;
            if (!current->children[bit]) {
                current->children[bit] = std::make_unique<TrieNode>();
            }
            current = current->children[bit].get();
            current->count++;
        }
    }
    
    void remove(int num) {
        TrieNode* current = root.get();
        for (int i = MAX_BITS; i >= 0; i--) {
            int bit = (num >> i) & 1;
            if (current->children[bit]) {
                current = current->children[bit].get();
                current->count--;
            }
        }
    }
    
    // Find maximum XOR with given number
    int getMaxXOR(int num) {
        TrieNode* current = root.get();
        int maxXor = 0;
        
        for (int i = MAX_BITS; i >= 0; i--) {
            int bit = (num >> i) & 1;
            int oppositeBit = 1 - bit;
            
            // Try to go opposite direction for max XOR
            if (current->children[oppositeBit] && current->children[oppositeBit]->count > 0) {
                maxXor |= (1 << i);
                current = current->children[oppositeBit].get();
            } else if (current->children[bit] && current->children[bit]->count > 0) {
                current = current->children[bit].get();
            } else {
                break;
            }
        }
        return maxXor;
    }
};

int main() {
    Trie trie;
    
    // Basic operations
    trie.insert("apple");
    trie.insert("app");
    trie.insert("application");
    trie.insert("apply");
    
    std::cout << "Search 'apple': " << (trie.search("apple") ? "true" : "false") << std::endl;
    std::cout << "Search 'app': " << (trie.search("app") ? "true" : "false") << std::endl;
    std::cout << "Search 'appl': " << (trie.search("appl") ? "true" : "false") << std::endl;
    std::cout << "StartsWith 'app': " << (trie.startsWith("app") ? "true" : "false") << std::endl;
    std::cout << "Words with prefix 'app': " << trie.countWordsWithPrefix("app") << std::endl;
    
    std::cout << "All words with prefix 'app': ";
    for (const auto& word : trie.getWordsWithPrefix("app")) {
        std::cout << word << " ";
    }
    std::cout << std::endl;
    
    // XOR Trie demo
    XORTrie xorTrie;
    std::vector<int> nums = {3, 10, 5, 25, 2, 8};
    for (int num : nums) {
        xorTrie.insert(num);
    }
    
    int maxXor = 0;
    for (int num : nums) {
        maxXor = std::max(maxXor, xorTrie.getMaxXOR(num));
    }
    std::cout << "Maximum XOR in array: " << maxXor << std::endl; // Should be 28 (5 XOR 25)
    
    return 0;
}
